tsdb: Delete blocks atomically; Remove tmp blocks on start; Added test.

bwplotka opened the pull request:
Reason: When panic / `SEGFault` happens during deletion TSDB, can't start gracefully ): 

## Changes:

* Rename dir when deleting
* Ignoring blocks with broken meta.json on start (we do that on reload)
* Compactor writes <ulid>.tmp-for-creation blocks instead of just .tmp
* Delete tmp-for-creation and tmp-for-deletion blocks during DB open.

cc @brian-brazil @brancz @metalmatze @codesome 

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

kakkoyun commented on tsdb/db.go:
```diff
@@ -1016,7 +1017,7 @@ func DefaultBlocksToDelete(db *DB) BlocksToDeleteFunc {
 }
 
-// deletableBlocks returns all blocks past retention policy.
+// deletableBlocks returns all currently loaded blocks past retention policy or already compacted into new block.
```
```suggestion
// deletableBlocks returns all currently loaded blocks past retention policy or already compacted into a new block.
```

squat commented on tsdb/db.go:
```diff
@@ -1105,7 +1106,13 @@ func (db *DB) deleteBlocks(blocks map[ulid.ULID]*Block) error {
+		// Rename atomically to avoid partial block when process crashing during this moment.
+		tmpToDelete := filepath.Join(db.dir, fmt.Sprintf("%s.tmp-to-delete", ulid))
+		if err := os.Rename(filepath.Join(db.dir, ulid.String()), tmpToDelete); err != nil {
+			return errors.Wrapf(err, "atomic rename for delete obsolete block %s", ulid)
```
```suggestion
			return errors.Wrapf(err, "atomic rename for deleting obsolete block %s", ulid)
```

squat commented on tsdb/db.go:
```diff
@@ -1105,7 +1106,13 @@ func (db *DB) deleteBlocks(blocks map[ulid.ULID]*Block) error {
 		}
-		if err := os.RemoveAll(filepath.Join(db.dir, ulid.String())); err != nil {
+
+		// Rename atomically to avoid partial block when process crashing during this moment.
```
```suggestion
		// Rename atomically to avoid partial block when process is crashing during this moment.
```

bwplotka commented:
Should we aim for deleting those on start in this PR? @squat @kakkoyun ?

kakkoyun commented:
@bwplotka I'd say yes. Since it's low-hanging fruit.

roidelapluie commented:
We should do this in 2 passes:

1. rename all blocks to tmp-to-delete

2. delete the tmp-to-delete

WDYT?

roidelapluie commented:
The alternative to remove the meta file first seems great too.

krasi-georgiev commented on tsdb/db.go:
```diff
@@ -1105,7 +1106,13 @@ func (db *DB) deleteBlocks(blocks map[ulid.ULID]*Block) error {
 		}
-		if err := os.RemoveAll(filepath.Join(db.dir, ulid.String())); err != nil {
+
+		// Rename atomically to avoid partial block when process crashing during this moment.
```
use `fileutil.Replace` ?

krasi-georgiev replied:
can't use `fileutil.Replace` ?

krasi-georgiev commented on tsdb/db.go:
```diff
@@ -549,6 +549,7 @@ func validateOpts(opts *Options, rngs []int64) (*Options, []int64) {
 	return opts, rngs
 }
 
+// TODO(bwplotka): Detect .tmp and .tmp-to-delete dirs and remove on start.
```
Why a todo? The PR is small enough to add here. Then it will be within the same context why these tmp might folders exist anyway.

krasi-georgiev commented:
yes to deleting  the tmp folders on start up.

bwplotka commented:
> We should do this in 2 passes:
> 
> rename all blocks to tmp-to-delete
> 
> delete the tmp-to-delete
> 
> WDYT?

Why it matters @roidelapluie 

I will explore removing meta.json then and assuming such block ready to be deleted

roidelapluie commented:
> > We should do this in 2 passes:
> > rename all blocks to tmp-to-delete
> > delete the tmp-to-delete
> > WDYT?
> 
> Why it matters @roidelapluie
> 
> I will explore removing meta.json then and assuming such block ready to be deleted

You protect with a crash in the middle of the deletes, but a crash at that time would not guarantee that all the deleted blocks would be deleted, or unseen.

bwplotka commented:
Ok, addressed all comments and changed a bit. All changes are in updated [description](https://github.com/prometheus/prometheus/pull/7772#issue-465470101)

Can you re-PTAL? @pracucci @kakkoyun @roidelapluie @krasi-georgiev ? (:

roidelapluie commented:
Oh I know why I wanted this to first rename all then delete:

If it crashed while deleting, if that is during compaction, Prometheus still would not restart unless you have the vertical merge enabled. Is this function called after compaction too?

bwplotka commented:
> If it crashed while deleting, if that is during compaction, Prometheus still would not restart unless you have the vertical merge enabled. Is this function called after compaction too?

Ah good point. Need to look up all deletions really. Including done by Delete API if full block is to remove etc.

bwplotka commented:
Ok, the delete on reload function is used for compaction as well - so kind of same delete path is used. 

However the crash during that path is fine - as long as deletion is atomic. We handed those overlaps just fine thanks to reload functionality of deduping those if parent matches:

> // By creating blocks with their parents, we can pick up the deletion where it left off during a crash

https://github.com/prometheus/prometheus/blob/9c12b0146897bd8a4487e7155e0e4783e6fe036d/tsdb/db.go#L927

So we are safe on this case (:
